MultipartUpload with uploadPartCopy not working

I'm having an issue running a MultipartUpload using uploadPartCopy that works with S3 but fails with S3Mock. I'm getting this error **"One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part's entity tag."**

I've attached the full test code, output running against s3, output running against s3 mock, error traceback, and docker container logs below.

I think the most relevant part of the logs is from the docker container:

**Mapped to com.adobe.testing.s3mock.FileStoreController#copyObject**

```
PUT "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?partNumber=1&uploadId=21eba32e-7ad6-45fd-8e27-28ae7ed423b6", parameters={masked}
Mapped to com.adobe.testing.s3mock.FileStoreController#copyObject(String, ObjectRef, MetadataDirective, String, String, HttpServletRequest)
Using 'application/xml', given [application/x-www-form-urlencoded, application/xml] and supported [application/xml]
Writing [com.adobe.testing.s3mock.dto.CopyObjectResult@52c5661c]
Completed 200 OK
PUT "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?partNumber=2&uploadId=21eba32e-7ad6-45fd-8e27-28ae7ed423b6", parameters={masked}
Mapped to com.adobe.testing.s3mock.FileStoreController#copyObject(String, ObjectRef, MetadataDirective, String, String, HttpServletRequest)
Using 'application/xml', given [application/x-www-form-urlencoded, application/xml] and supported [application/xml]
Writing [com.adobe.testing.s3mock.dto.CopyObjectResult@7a1e78f8]
Completed 200 OK
```

My best guess is that this should be using `copyObjectPart` not `copyObject` (or another function that maybe isn't implemented?). My request is uploading the entire object as a part, so there is no `COPY_SOURCE_RANGE` header. 

https://github.com/adobe/S3Mock/blob/4cb27728a9c60e48754fc152ecf08b41ae1345fa/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java#L768-L775

https://github.com/adobe/S3Mock/blob/4cb27728a9c60e48754fc152ecf08b41ae1345fa/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java#L868-L874

<details>
<summary>
Test Code
</summary>

```java
package com.whatever.whatever;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutionException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.model.CompletedPart;
import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.core.async.AsyncResponseTransformer;

import org.junit.jupiter.api.Test;

public class DebugS3Mock {

    S3AsyncClient client = S3AsyncClient.builder()
        .serviceConfiguration(x -> x.pathStyleAccessEnabled(true))
        // .endpointOverride(URI.create("http://localhost:9090"))
        // .credentialsProvider(AnonymousCredentialsProvider.create())
        // .region(Region.US_EAST_1)
        .build();

    private Path getAbsoluteTestPath(String relativePath) {
        String absolutePath = this.getClass().getClassLoader().getResource(relativePath).getFile();
        return Paths.get(absolutePath);
    }

    @Test
    public void testS3Mock_MultiCopyGet() throws InterruptedException, ExecutionException, IOException {
        UUID uuid = UUID.randomUUID();

        System.out.println("Creating sourceBucket...");
        var sourceBucket = "test-bucket1-" +  uuid.toString();;
        client.createBucket(r -> r.bucket(sourceBucket)).get();
        System.out.println("Created sourceBucket: " + sourceBucket);

        System.out.println("Uploading test data to sourceBucket...");
        String testFolder = "multipart-test-data";
        String[] sourceKeys = { "test-file.part0", "test-file.part1" };
        for (String key : sourceKeys) {
            Path path = getAbsoluteTestPath(testFolder + "/" + key);
            client.putObject(r -> r.bucket(sourceBucket).key(key), AsyncRequestBody.fromFile(path)).get();
        }
        System.out.println("Finished uploading " + sourceKeys.toString());

        System.out.println("Creating destinationBucket...");
        var destinationBucket = "test-bucket2-" +  uuid.toString();;
        client.createBucket(r -> r.bucket(destinationBucket)).get();
        System.out.println("Created destinationBucket: " + sourceBucket);

        System.out.println("Starting create multipart upload...");
        var destinationKey = "test-file";
        var createMultipartUploadResponse = client.createMultipartUpload(r -> r.bucket(destinationBucket).key(destinationKey)).get();
        var uploadId = createMultipartUploadResponse.uploadId();
        System.out.println("Created multipart upload with upload id: " + uploadId);

        System.out.println("Uploading individual parts...");
        List<CompletedPart> parts = new ArrayList<CompletedPart>();
        for (int i = 0; i < sourceKeys.length; i++) {
            var sourceKey = sourceKeys[i];
            var partNumber = i + 1;
            
            System.out.println("Uploading part " + partNumber + " (" + sourceKey + ")...");
            var uploadPartCopyResponse = client.uploadPartCopy(t -> t.sourceBucket(sourceBucket)
                .sourceKey(sourceKey)
                .uploadId(uploadId)
                .destinationBucket(destinationBucket)
                .destinationKey(destinationKey)
                .partNumber(partNumber)).get();
            var eTag = uploadPartCopyResponse.copyPartResult().eTag();
            System.out.println("Uploaded with eTag: " + eTag);

            var part = CompletedPart.builder().eTag(eTag).partNumber(partNumber).build();
            parts.add(part);
        }

        System.out.println("Completing multipart upload...");
        var completeMultipartUploadResponse = client.completeMultipartUpload(r -> {
            r.bucket(destinationBucket)
                    .key(destinationKey)
                    .uploadId(uploadId)
                    .multipartUpload(s -> s.parts(parts));
        }).get();
        System.out.println("Completed multipart upload...");

        System.out.println("Checking if local (combined) test-file is the same as the remote test-file in destination bucket...");
        var localKey = "test-file"; // local file that is combined version of part0 and part1
        var localBytes = Files.readAllBytes(getAbsoluteTestPath(testFolder + "/" + localKey));
        var remoteBytes = client
            .getObject(r -> r.bucket(destinationBucket).key(destinationKey), AsyncResponseTransformer.toBytes()).get()
            .asByteArray();
        
        if (Arrays.equals(localBytes, remoteBytes)) {
            System.out.println("its the same!");
        }
        
        System.out.println("Cleaning up...");
        for (String sourceKey : sourceKeys) {
            client.deleteObject(r -> r.bucket(sourceBucket).key(sourceKey)).get();
        }
        client.deleteObject(r -> r.bucket(destinationBucket).key(destinationKey)).get();
        client.deleteBucket(r -> r.bucket(sourceBucket)).get();
        client.deleteBucket(r -> r.bucket(destinationBucket)).get();

        System.out.println("All Done!");

        // assertEquals(localBytes, remoteBytes);
    }

}
```
</details>

<details>
<summary>
Output Log hit against S3
</summary>

>**Creating sourceBucket...**
2022-12-30 06:05:46 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:46 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Created sourceBucket: test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8**
**Uploading test data to sourceBucket...**
2022-12-30 06:05:46 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8/test-file.part0, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, Expect, User-Agent], queryParameters=[])
2022-12-30 06:05:47 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
2022-12-30 06:05:47 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8/test-file.part1, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, Expect, User-Agent], queryParameters=[])
2022-12-30 06:05:47 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Finished uploading [Ljava.lang.String;@767a014e**
**Creating destinationBucket...**
2022-12-30 06:05:47 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:47 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Created destinationBucket: test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8**
**Starting create multipart upload...**
2022-12-30 06:05:47 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=POST, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, User-Agent], queryParameters=[uploads])
2022-12-30 06:05:48 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Created multipart upload with upload id:** NHF6wyepA4RJPrU1YyYByHRBnEuOaJ_7GElzBiBX8R_q84sGZWz.33OSvS1FlZOoZmRD1YfG8MqdP8StacAxKSTonIHfz7c.87pb.i90f1LmHK0p2VGLuVkzFcnVEO.AP0l7AsKE.wtOG2EEwolRjrA6jmsNqTGZc4RNK9U8AN4-
**Uploading individual parts...**
**Uploading part 1 (test-file.part0)...**
2022-12-30 06:05:48 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, User-Agent, x-amz-copy-source], queryParameters=[partNumber, uploadId])
2022-12-30 06:05:48 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Uploaded with eTag: "79224696d1dffcbe178a01e5289899ee"**
**Uploading part 2 (test-file.part1)...**
2022-12-30 06:05:48 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, User-Agent, x-amz-copy-source], queryParameters=[partNumber, uploadId])
2022-12-30 06:05:48 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Uploaded with eTag: "75fb978cf631239e26a92707be35d512"**
**Completing multipart upload...**
2022-12-30 06:05:48 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=POST, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, User-Agent], queryParameters=[uploadId])
2022-12-30 06:05:48 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**Completed multipart upload...**
**Checking if local (combined) test-file is the same as the remote test-file in destination bucket...**
2022-12-30 06:05:48 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=GET, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, User-Agent, x-amz-te], queryParameters=[])
2022-12-30 06:05:48 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
**its the same!**
**Cleaning up...**
2022-12-30 06:05:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=DELETE, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8/test-file.part0, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 204
2022-12-30 06:05:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=DELETE, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8/test-file.part1, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 204
2022-12-30 06:05:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=DELETE, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8/test-file, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 204
2022-12-30 06:05:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=DELETE, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket1-5625af65-c46c-4613-bdc2-30db85e040f8, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 204
2022-12-30 06:05:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=DELETE, protocol=https, host=s3.amazonaws.com, port=443, encodedPath=/test-bucket2-5625af65-c46c-4613-bdc2-30db85e040f8, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:05:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 204
**All Done!**
</details>


<details>
<summary>
Output Log hit against S3Mock
</summary>

>Creating sourceBucket...
2022-12-30 06:15:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket1-e36f5e29-b702-408a-8999-45600298e952, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:15:49 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
Created sourceBucket: test-bucket1-e36f5e29-b702-408a-8999-45600298e952
Uploading test data to sourceBucket...
2022-12-30 06:15:49 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket1-e36f5e29-b702-408a-8999-45600298e952/test-file.part0, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, Expect, User-Agent], queryParameters=[])
2022-12-30 06:15:50 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
2022-12-30 06:15:50 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket1-e36f5e29-b702-408a-8999-45600298e952/test-file.part1, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, Expect, User-Agent], queryParameters=[])
2022-12-30 06:15:50 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
Finished uploading [Ljava.lang.String;@6ffa56fa
Creating destinationBucket...
2022-12-30 06:15:50 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket2-e36f5e29-b702-408a-8999-45600298e952, headers=[amz-sdk-invocation-id, User-Agent], queryParameters=[])
2022-12-30 06:15:50 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
Created destinationBucket: test-bucket1-e36f5e29-b702-408a-8999-45600298e952
Starting create multipart upload...
2022-12-30 06:15:50 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=POST, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file, headers=[amz-sdk-invocation-id, Content-Length, Content-Type, User-Agent], queryParameters=[uploads])
2022-12-30 06:15:50 [aws-java-sdk-NettyEventLoop-1-2] DEBUG software.amazon.awssdk.request:85 - Received successful response: 200
Created multipart upload with upload id: 21eba32e-7ad6-45fd-8e27-28ae7ed423b6
Uploading individual parts...
Uploading part 1 (test-file.part0)...
2022-12-30 06:15:50 [main] DEBUG software.amazon.awssdk.request:85 - Sending Request: DefaultSdkHttpFullRequest(httpMethod=PUT, protocol=http, host=localhost, port=9090, encodedPath=/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file, headers=[amz-sdk-invocation-id, User-Agent, x-amz-copy-source], queryParameters=[partNumber, uploadId])
</details>

<details>
<summary>
Error Traceback
</summary>

java.util.concurrent.ExecutionException: software.amazon.awssdk.services.s3.model.S3Exception: One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part's entity tag. (Service: S3, Status Code: 400, Request ID: , Extended Request ID: null)
 at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
 at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
 at com.amazon.sample.DebugS3Mock.testS3Mock_MultiCopyGet(DebugS3Mock.java:94)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:566)
 at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
 at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
 at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
 at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
 at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
 at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
 at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
 at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
 at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
 at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
 at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
 at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
 at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
 at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
 at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
 at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
 at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
 at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
 at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
 at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
 at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
 at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
 at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
 at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
 at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
 at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
 at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
 at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
 at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
 at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
 at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
 at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
 at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
 at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
 at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
 at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:84)
 at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
 at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
 at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: software.amazon.awssdk.services.s3.model.S3Exception: One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part's entity tag. (Service: S3, Status Code: 400, Request ID: , Extended Request ID: null)
 at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleErrorResponse(AwsXmlPredicatedResponseHandler.java:156)
 at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handleResponse(AwsXmlPredicatedResponseHandler.java:106)
 at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:84)
 at software.amazon.awssdk.protocols.xml.internal.unmarshall.AwsXmlPredicatedResponseHandler.handle(AwsXmlPredicatedResponseHandler.java:42)
 at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$successTransformationResponseHandler$6(BaseClientHandler.java:234)
 at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:89)
 at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
 at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
 at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
 at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:132)
 at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.runAndLogError(ResponseHandler.java:181)
 at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler.access$500(ResponseHandler.java:71)
 at software.amazon.awssdk.http.nio.netty.internal.ResponseHandler$PublisherAdapter$1.onComplete(ResponseHandler.java:298)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.complete(HandlerPublisher.java:447)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HandlerPublisher.handlerRemoved(HandlerPublisher.java:435)
 at io.netty.channel.AbstractChannelHandlerContext.callHandlerRemoved(AbstractChannelHandlerContext.java:946)
 at io.netty.channel.DefaultChannelPipeline.callHandlerRemoved0(DefaultChannelPipeline.java:637)
 at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:477)
 at io.netty.channel.DefaultChannelPipeline.remove(DefaultChannelPipeline.java:423)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.removeHandlerIfActive(HttpStreamsHandler.java:362)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.handleReadHttpContent(HttpStreamsHandler.java:223)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsHandler.channelRead(HttpStreamsHandler.java:199)
 at software.amazon.awssdk.http.nio.netty.internal.nrs.HttpStreamsClientHandler.channelRead(HttpStreamsClientHandler.java:173)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at software.amazon.awssdk.http.nio.netty.internal.LastHttpContentHandler.channelRead(LastHttpContentHandler.java:43)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
 at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
 at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
 at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
 at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
 at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
 at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
 at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
 at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
 at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
 at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
 at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
 at java.base/java.lang.Thread.run(Thread.java:829)
</details>


<details>
<summary>
Docker Run Debug Logs
</summary>

```
2021-12-30 06:15:24.008  INFO 1 --- [           main] c.a.testing.s3mock.S3MockApplication     : Started S3MockApplication in 2.306 seconds (JVM running for 2.757)
2021-12-30 06:15:24.009 DEBUG 1 --- [           main] o.s.b.a.ApplicationAvailabilityBean      : Application availability state LivenessState changed to CORRECT
2021-12-30 06:15:24.010 DEBUG 1 --- [           main] o.s.b.a.ApplicationAvailabilityBean      : Application availability state ReadinessState changed to ACCEPTING_TRAFFIC
2021-12-30 06:15:49.757 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket1-e36f5e29-b702-408a-8999-45600298e952", parameters={}
2021-12-30 06:15:49.766 DEBUG 1 --- [tp1940445711-51] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#createBucket(String)
2021-12-30 06:15:49.810 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/x-www-form-urlencoded', given [application/x-www-form-urlencoded, application/xml] and supported [text/plain, */*, text/plain, */*, application/json, application/*+json, application/json, application/*+json, application/xml, application/x-www-form-urlencoded, application/octet-stream, application/xml;charset=UTF-8, text/xml;charset=UTF-8, application/*+xml;charset=UTF-8]
2021-12-30 06:15:49.810 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Nothing to write: null body
2021-12-30 06:15:49.811 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:49.874 DEBUG 1 --- [tp1940445711-39] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket1-e36f5e29-b702-408a-8999-45600298e952/test-file.part0", parameters={}
2021-12-30 06:15:49.875 DEBUG 1 --- [tp1940445711-39] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#putObject(String, String, String, List, String, String, String, HttpServletRequest)
2021-12-30 06:15:50.211 DEBUG 1 --- [tp1940445711-39] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/x-www-form-urlencoded', given [application/x-www-form-urlencoded, application/xml] and supported [text/plain, */*, text/plain, */*, application/json, application/*+json, application/json, application/*+json, application/xml, application/x-www-form-urlencoded, application/octet-stream, application/xml;charset=UTF-8, text/xml;charset=UTF-8, application/*+xml;charset=UTF-8]
2021-12-30 06:15:50.212 DEBUG 1 --- [tp1940445711-39] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Nothing to write: null body
2021-12-30 06:15:50.212 DEBUG 1 --- [tp1940445711-39] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.228 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket1-e36f5e29-b702-408a-8999-45600298e952/test-file.part1", parameters={}
2021-12-30 06:15:50.229 DEBUG 1 --- [tp1940445711-49] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#putObject(String, String, String, List, String, String, String, HttpServletRequest)
2021-12-30 06:15:50.352 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/x-www-form-urlencoded', given [application/x-www-form-urlencoded, application/xml] and supported [text/plain, */*, text/plain, */*, application/json, application/*+json, application/json, application/*+json, application/xml, application/x-www-form-urlencoded, application/octet-stream, application/xml;charset=UTF-8, text/xml;charset=UTF-8, application/*+xml;charset=UTF-8]
2021-12-30 06:15:50.352 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Nothing to write: null body
2021-12-30 06:15:50.352 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.360 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952", parameters={}
2021-12-30 06:15:50.361 DEBUG 1 --- [tp1940445711-51] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#createBucket(String)
2021-12-30 06:15:50.362 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/x-www-form-urlencoded', given [application/x-www-form-urlencoded, application/xml] and supported [text/plain, */*, text/plain, */*, application/json, application/*+json, application/json, application/*+json, application/xml, application/x-www-form-urlencoded, application/octet-stream, application/xml;charset=UTF-8, text/xml;charset=UTF-8, application/*+xml;charset=UTF-8]
2021-12-30 06:15:50.362 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Nothing to write: null body
2021-12-30 06:15:50.362 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.383 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : POST "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?uploads", parameters={masked}
2021-12-30 06:15:50.384 DEBUG 1 --- [tp1940445711-49] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#initiateMultipartUpload(String, String, String, HttpServletRequest)
2021-12-30 06:15:50.387 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/xml', given [application/x-www-form-urlencoded, application/xml] and supported [application/xml]
2021-12-30 06:15:50.391 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Writing [com.adobe.testing.s3mock.dto.InitiateMultipartUploadResult@d81937f]
2021-12-30 06:15:50.407 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.441 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?partNumber=1&uploadId=21eba32e-7ad6-45fd-8e27-28ae7ed423b6", parameters={masked}
2021-12-30 06:15:50.442 DEBUG 1 --- [tp1940445711-51] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#copyObject(String, ObjectRef, MetadataDirective, String, String, HttpServletRequest)
2021-12-30 06:15:50.522 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/xml', given [application/x-www-form-urlencoded, application/xml] and supported [application/xml]
2021-12-30 06:15:50.526 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Writing [com.adobe.testing.s3mock.dto.CopyObjectResult@52c5661c]
2021-12-30 06:15:50.528 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.563 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : PUT "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?partNumber=2&uploadId=21eba32e-7ad6-45fd-8e27-28ae7ed423b6", parameters={masked}
2021-12-30 06:15:50.564 DEBUG 1 --- [tp1940445711-49] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#copyObject(String, ObjectRef, MetadataDirective, String, String, HttpServletRequest)
2021-12-30 06:15:50.644 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Using 'application/xml', given [application/x-www-form-urlencoded, application/xml] and supported [application/xml]
2021-12-30 06:15:50.644 DEBUG 1 --- [tp1940445711-49] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Writing [com.adobe.testing.s3mock.dto.CopyObjectResult@7a1e78f8]
2021-12-30 06:15:50.645 DEBUG 1 --- [tp1940445711-49] o.s.web.servlet.DispatcherServlet        : Completed 200 OK
2021-12-30 06:15:50.660 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : POST "/test-bucket2-e36f5e29-b702-408a-8999-45600298e952/test-file?uploadId=21eba32e-7ad6-45fd-8e27-28ae7ed423b6", parameters={masked}
2021-12-30 06:15:50.660 DEBUG 1 --- [tp1940445711-51] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to com.adobe.testing.s3mock.FileStoreController#completeMultipartUpload(String, String, String, String, CompleteMultipartUploadRequest, HttpServletRequest)
2021-12-30 06:15:50.704 DEBUG 1 --- [tp1940445711-51] m.m.a.RequestResponseBodyMethodProcessor : Read "application/xml;charset=UTF-8" to [com.adobe.testing.s3mock.dto.CompleteMultipartUploadRequest@339d87f7]
2021-12-30 06:15:50.709  INFO 1 --- [tp1940445711-51] c.a.t.s.util.S3MockExceptionHandler      : Responding with status 400: One or more of the specified parts could not be found. The part might not have been uploaded, or the specified entity tag might not have matched the part's entity tag.
2021-12-30 06:15:50.710 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Found 'Content-Type:application/xml' in response
2021-12-30 06:15:50.711 DEBUG 1 --- [tp1940445711-51] o.s.w.s.m.m.a.HttpEntityMethodProcessor  : Writing [com.adobe.testing.s3mock.dto.ErrorResponse@674babdf]
2021-12-30 06:15:50.712 DEBUG 1 --- [tp1940445711-51] o.s.web.servlet.DispatcherServlet        : Completed 400 BAD_REQUEST
```
</details>

--- a/integration-tests/src/test/java/com/adobe/testing/s3mock/its/MultiPartUploadIT.java
+++ b/integration-tests/src/test/java/com/adobe/testing/s3mock/its/MultiPartUploadIT.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright 2017-2021 Adobe.
+ *  Copyright 2017-2022 Adobe.
  *
  *  Licensed under the Apache License, Version 2.0 (the "License");
  *  you may not use this file except in compliance with the License.
@@ -317,6 +317,78 @@ public class MultiPartUploadIT extends S3TestBase {
             "Object contents doesn't match")
         .isEqualTo(concatByteArrays(randomBytes1, randomBytes3));
 
+  }
+
+  /**
+   * Upload two objects, copy as parts without length, complete multipart.
+   */
+  @Test
+  void shouldCopyPartsAndComplete() throws IOException {
+    //Initiate upload in BUCKET_NAME_2
+    s3Client.createBucket(BUCKET_NAME_2);
+    String multipartUploadKey = UUID.randomUUID().toString();
+    final InitiateMultipartUploadResult initiateMultipartUploadResult = s3Client
+        .initiateMultipartUpload(
+            new InitiateMultipartUploadRequest(BUCKET_NAME_2, multipartUploadKey));
+    final String uploadId = initiateMultipartUploadResult.getUploadId();
+    List<PartETag> parts = new ArrayList<>();
+
+    //bucket for test data
+    s3Client.createBucket(BUCKET_NAME);
+
+    //create two objects, initiate copy part with full object length
+    String[] sourceKeys = {UUID.randomUUID().toString(), UUID.randomUUID().toString()};
+    List<byte[]> allRandomBytes = new ArrayList<>();
+    for (int i = 0; i < sourceKeys.length; i++) {
+      String key = sourceKeys[i];
+      int partNumber = i + 1;
+      byte[] randomBytes = createRandomBytes();
+      ObjectMetadata metadata1 = new ObjectMetadata();
+      metadata1.setContentLength(randomBytes.length);
+      PutObjectResult putObjectResult = //ignore result
+          s3Client.putObject(
+              new PutObjectRequest(BUCKET_NAME, key, new ByteArrayInputStream(randomBytes),
+                  metadata1));
+      CopyPartRequest request = new CopyPartRequest()
+          .withPartNumber(partNumber)
+          .withUploadId(uploadId)
+          .withDestinationBucketName(BUCKET_NAME_2)
+          .withDestinationKey(multipartUploadKey)
+          .withSourceKey(key)
+          .withSourceBucketName(BUCKET_NAME);
+      CopyPartResult result = s3Client.copyPart(request);
+      String etag = result.getETag();
+      PartETag partETag = new PartETag(partNumber, etag);
+      parts.add(partETag);
+      allRandomBytes.add(randomBytes);
+    }
+
+    assertThat(allRandomBytes).hasSize(2);
+
+    // Complete with parts
+    final CompleteMultipartUploadResult result = s3Client.completeMultipartUpload(
+        new CompleteMultipartUploadRequest(BUCKET_NAME_2, multipartUploadKey, uploadId, parts));
+
+    // Verify parts
+    S3Object object = s3Client.getObject(BUCKET_NAME_2, multipartUploadKey);
+
+    byte[] allMd5s = ArrayUtils.addAll(
+        DigestUtils.md5(allRandomBytes.get(0)),
+        DigestUtils.md5(allRandomBytes.get(1))
+    );
+
+    // verify etag
+    assertThat(result.getETag()).as("etag doesn't match.")
+        .isEqualTo(DigestUtils.md5Hex(allMd5s) + "-2");
+
+    // verify content size
+    assertThat(object.getObjectMetadata().getContentLength()).as("Content length doesn't match")
+        .isEqualTo((long) allRandomBytes.get(0).length + allRandomBytes.get(1).length);
+
+    // verify contents
+    assertThat(readStreamIntoByteArray(object.getObjectContent())).as(
+            "Object contents doesn't match")
+        .isEqualTo(concatByteArrays(allRandomBytes.get(0), allRandomBytes.get(1)));
   }
 
   /**

--- a/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright 2017-2021 Adobe.
+ *  Copyright 2017-2022 Adobe.
  *
  *  Licensed under the Apache License, Version 2.0 (the "License");
  *  you may not use this file except in compliance with the License.
@@ -25,6 +25,17 @@ import static com.adobe.testing.s3mock.util.AwsHttpHeaders.NOT_COPY_SOURCE_RANGE
 import static com.adobe.testing.s3mock.util.AwsHttpHeaders.RANGE;
 import static com.adobe.testing.s3mock.util.AwsHttpHeaders.SERVER_SIDE_ENCRYPTION;
 import static com.adobe.testing.s3mock.util.AwsHttpHeaders.SERVER_SIDE_ENCRYPTION_AWS_KMS_KEYID;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.CONTINUATION_TOKEN;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.DELETE;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.ENCODING_TYPE;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.LIST_TYPE_V2;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.MAX_KEYS;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.NOT_UPLOAD_ID;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.PART_NUMBER;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.START_AFTER;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.TAGGING;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.UPLOADS;
+import static com.adobe.testing.s3mock.util.AwsHttpParameters.UPLOAD_ID;
 import static com.adobe.testing.s3mock.util.MetadataUtil.createUserMetadataHeaders;
 import static com.adobe.testing.s3mock.util.MetadataUtil.getUserMetadata;
 import static com.adobe.testing.s3mock.util.StringEncoding.decode;
@@ -46,6 +57,7 @@ import static org.springframework.http.HttpStatus.NOT_MODIFIED;
 import static org.springframework.http.HttpStatus.PARTIAL_CONTENT;
 import static org.springframework.http.HttpStatus.PRECONDITION_FAILED;
 import static org.springframework.http.HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE;
+import static org.springframework.http.MediaType.APPLICATION_XML_VALUE;
 
 import com.adobe.testing.s3mock.domain.Bucket;
 import com.adobe.testing.s3mock.domain.BucketContents;
@@ -119,8 +131,6 @@ import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBo
 @RestController
 class FileStoreController {
 
-  private static final String ANY = "*";
-
   private static final String RANGES_BYTES = "bytes";
 
   private static final String STREAMING_AWS_4_HMAC_SHA_256_PAYLOAD =
@@ -155,15 +165,19 @@ class FileStoreController {
   // /
   //================================================================================================
 
-
   /**
    * List all existing buckets.
    * <p>https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListBuckets.html</p>
    *
    * @return List of all Buckets
    */
-  @RequestMapping(value = "/", method = RequestMethod.GET, produces = {
-      "application/xml"})
+  @RequestMapping(
+      value = "/",
+      method = RequestMethod.GET,
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<ListAllMyBucketsResult> listBuckets() {
     return ResponseEntity.ok(new ListAllMyBucketsResult(TEST_OWNER, fileStore.listBuckets()));
   }
@@ -171,7 +185,6 @@ class FileStoreController {
   //================================================================================================
   // /{bucketName:.+}
   //================================================================================================
-
 
   /**
    * Create a bucket.
@@ -181,7 +194,10 @@ class FileStoreController {
    *
    * @return 200 OK if creation was successful.
    */
-  @RequestMapping(value = "/{bucketName}", method = RequestMethod.PUT)
+  @RequestMapping(
+      value = "/{bucketName}",
+      method = RequestMethod.PUT
+  )
   public ResponseEntity<String> createBucket(@PathVariable final String bucketName) {
     try {
       fileStore.createBucket(bucketName);
@@ -200,7 +216,10 @@ class FileStoreController {
    *
    * @return 200 if it exists; 404 if not found.
    */
-  @RequestMapping(value = "/{bucketName}", method = RequestMethod.HEAD)
+  @RequestMapping(
+      value = "/{bucketName}",
+      method = RequestMethod.HEAD
+  )
   public ResponseEntity<Void> headBucket(@PathVariable final String bucketName) {
     if (fileStore.doesBucketExist(bucketName)) {
       return ResponseEntity.ok().build();
@@ -217,7 +236,10 @@ class FileStoreController {
    *
    * @return 204 if Bucket was deleted; 404 if not found
    */
-  @RequestMapping(value = "/{bucketName}", method = RequestMethod.DELETE)
+  @RequestMapping(
+      value = "/{bucketName}",
+      method = RequestMethod.DELETE
+  )
   public ResponseEntity<String> deleteBucket(@PathVariable final String bucketName) {
     verifyBucketExistence(bucketName);
 
@@ -247,7 +269,7 @@ class FileStoreController {
    *
    * @param bucketName {@link String} set bucket name
    * @param prefix {@link String} find object names they starts with prefix
-   * @param encodingtype whether to use URL encoding (encodingtype="url") or not
+   * @param encodingType whether to use URL encoding (encodingtype="url") or not
    *
    * @return {@link ListBucketResult} a list of objects in Bucket
    * @deprecated Long since replaced by ListObjectsV2, {@see #listObjectsInsideBucketV2}
@@ -255,27 +277,30 @@ class FileStoreController {
   @RequestMapping(
       value = "/{bucketName}",
       method = RequestMethod.GET,
-      produces = {"application/xml"})
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   @Deprecated
   public ResponseEntity<ListBucketResult> listObjectsInsideBucket(
       @PathVariable final String bucketName,
       @RequestParam(required = false) final String prefix,
       @RequestParam(required = false) final String delimiter,
       @RequestParam(required = false) final String marker,
-      @RequestParam(name = "encoding-type", required = false) final String encodingtype,
-      @RequestParam(name = "max-keys", defaultValue = "1000",
+      @RequestParam(name = ENCODING_TYPE, required = false) final String encodingType,
+      @RequestParam(name = MAX_KEYS, defaultValue = "1000",
           required = false) final Integer maxKeys) {
     verifyBucketExistence(bucketName);
     if (maxKeys < 0) {
       throw new S3Exception(HttpStatus.BAD_REQUEST.value(), "InvalidRequest",
           "maxKeys should be non-negative");
     }
-    if (isNotEmpty(encodingtype) && !"url".equals(encodingtype)) {
+    if (isNotEmpty(encodingType) && !"url".equals(encodingType)) {
       throw new S3Exception(HttpStatus.BAD_REQUEST.value(), "InvalidRequest",
           "encodingtype can only be none or 'url'");
     }
 
-    final boolean useUrlEncoding = Objects.equals("url", encodingtype);
+    final boolean useUrlEncoding = Objects.equals("url", encodingType);
 
     try {
       List<BucketContents> contents = getBucketContents(bucketName, prefix);
@@ -305,7 +330,7 @@ class FileStoreController {
 
       return ResponseEntity.ok(
           new ListBucketResult(bucketName, returnPrefix, marker, maxKeys, isTruncated,
-              encodingtype, nextMarker, contents, returnCommonPrefixes));
+              encodingType, nextMarker, contents, returnCommonPrefixes));
     } catch (final IOException e) {
       LOG.error("Object(s) could not retrieved from bucket {}", bucketName, e);
       return ResponseEntity.status(INTERNAL_SERVER_ERROR).build();
@@ -326,18 +351,24 @@ class FileStoreController {
    *
    * @return {@link ListBucketResultV2} a list of objects in Bucket
    */
-  @RequestMapping(value = "/{bucketName}", params = "list-type=2",
+  @RequestMapping(value = "/{bucketName}",
+      params = {
+          LIST_TYPE_V2
+      },
       method = RequestMethod.GET,
-      produces = {"application/xml"})
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<ListBucketResultV2> listObjectsInsideBucketV2(
       @PathVariable final String bucketName,
       @RequestParam(required = false) final String prefix,
       @RequestParam(required = false) final String delimiter,
-      @RequestParam(name = "encoding-type", required = false) final String encodingtype,
-      @RequestParam(name = "start-after", required = false) final String startAfter,
-      @RequestParam(name = "max-keys",
+      @RequestParam(name = ENCODING_TYPE, required = false) final String encodingtype,
+      @RequestParam(name = START_AFTER, required = false) final String startAfter,
+      @RequestParam(name = MAX_KEYS,
           defaultValue = "1000", required = false) final Integer maxKeys,
-      @RequestParam(name = "continuation-token", required = false) final String continuationToken) {
+      @RequestParam(name = CONTINUATION_TOKEN, required = false) final String continuationToken) {
     if (isNotEmpty(encodingtype) && !"url".equals(encodingtype)) {
       throw new S3Exception(HttpStatus.BAD_REQUEST.value(), "InvalidRequest",
           "encodingtype can only be none or 'url'");
@@ -412,9 +443,14 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/",
-      params = {"uploads"},
+      params = {
+          UPLOADS
+      },
       method = RequestMethod.GET,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<ListMultipartUploadsResult> listMultipartUploads(
       @PathVariable final String bucketName,
       @RequestParam(required = false) final String prefix,
@@ -458,9 +494,14 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName}",
-      params = "delete",
+      params = {
+          DELETE
+      },
       method = RequestMethod.POST,
-      produces = {"application/xml"})
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<BatchDeleteResponse> batchDeleteObjects(
       @PathVariable final String bucketName,
       @RequestBody final BatchDeleteRequest body) {
@@ -493,7 +534,8 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      method = RequestMethod.HEAD)
+      method = RequestMethod.HEAD
+  )
   public ResponseEntity<String> headObject(@PathVariable final String bucketName,
       final HttpServletRequest request) {
     verifyBucketExistence(bucketName);
@@ -527,7 +569,10 @@ class FileStoreController {
    *
    * @return ResponseEntity with Status Code 204 if object was successfully deleted.
    */
-  @RequestMapping(value = "/{bucketName:.+}/**", method = RequestMethod.DELETE)
+  @RequestMapping(
+      value = "/{bucketName:.+}/**",
+      method = RequestMethod.DELETE
+  )
   public ResponseEntity<String> deleteObject(@PathVariable final String bucketName,
       final HttpServletRequest request) {
     final String filename = filenameFrom(bucketName, request);
@@ -553,9 +598,14 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = {"uploadId"},
+      params = {
+          UPLOAD_ID
+      },
       method = RequestMethod.DELETE,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<Void> abortMultipartUpload(@PathVariable final String bucketName,
       @RequestParam final String uploadId,
       final HttpServletRequest request) {
@@ -579,7 +629,10 @@ class FileStoreController {
   @RequestMapping(
       value = "/{bucketName:.+}/**",
       method = RequestMethod.GET,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<StreamingResponseBody> getObject(@PathVariable final String bucketName,
       @RequestHeader(value = RANGE, required = false) final Range range,
       @RequestHeader(value = IF_MATCH, required = false) final List<String> match,
@@ -619,8 +672,11 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = "tagging",
-      method = RequestMethod.GET)
+      params = {
+          TAGGING
+      },
+      method = RequestMethod.GET
+  )
   public ResponseEntity<Tagging> getObjectTagging(@PathVariable final String bucketName,
       final HttpServletRequest request) {
     final String filename = filenameFrom(bucketName, request);
@@ -651,9 +707,14 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = {"uploadId"},
+      params = {
+          UPLOAD_ID
+      },
       method = RequestMethod.GET,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      }
+  )
   public ResponseEntity<ListPartsResult> multipartListParts(@PathVariable final String bucketName,
       @RequestParam final String uploadId,
       final HttpServletRequest request) {
@@ -674,8 +735,11 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = "tagging",
-      method = RequestMethod.PUT)
+      params = {
+          TAGGING
+      },
+      method = RequestMethod.PUT
+  )
   public ResponseEntity<String> putObjectTagging(@PathVariable final String bucketName,
       @RequestBody final Tagging body,
       final HttpServletRequest request) {
@@ -716,12 +780,16 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = {"uploadId", "partNumber"},
+      params = {
+          UPLOAD_ID,
+          PART_NUMBER
+      },
       headers = {
           NOT_COPY_SOURCE,
           NOT_COPY_SOURCE_RANGE
       },
-      method = RequestMethod.PUT)
+      method = RequestMethod.PUT
+  )
   public ResponseEntity<Void> putObjectPart(@PathVariable final String bucketName,
       @RequestParam final String uploadId,
       @RequestParam final String partNumber,
@@ -751,7 +819,7 @@ class FileStoreController {
    * <p>https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html</p>
    *
    * @param copySource References the Objects to be copied.
-   * @param copyRange Defines the byte range for this part.
+   * @param copyRange Defines the byte range for this part. Optional.
    * @param encryption The encryption type.
    * @param kmsKeyId The KMS encryption key id.
    * @param uploadId id of the upload. Has to match all other part's uploads.
@@ -764,11 +832,17 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{destinationBucket:.+}/**",
-      method = RequestMethod.PUT,
       headers = {
           COPY_SOURCE,
-          COPY_SOURCE_RANGE
-      }, produces = "application/xml")
+      },
+      params = {
+          UPLOAD_ID,
+          PART_NUMBER
+      },
+      method = RequestMethod.PUT,
+      produces = {
+          APPLICATION_XML_VALUE
+      })
   public ResponseEntity<CopyPartResult> copyObjectPart(
       @RequestHeader(value = COPY_SOURCE) final ObjectRef copySource,
       @RequestHeader(value = COPY_SOURCE_RANGE) final Range copyRange,
@@ -810,7 +884,10 @@ class FileStoreController {
    *
    * @throws IOException in case of an error on storing the object.
    */
-  @RequestMapping(value = "/{bucketName:.+}/**", method = RequestMethod.PUT)
+  @RequestMapping(
+      value = "/{bucketName:.+}/**",
+      method = RequestMethod.PUT
+  )
   public ResponseEntity<String> putObject(@PathVariable final String bucketName,
       @RequestHeader(value = SERVER_SIDE_ENCRYPTION, required = false) final String encryption,
       @RequestHeader(value = SERVER_SIDE_ENCRYPTION_AWS_KMS_KEYID, required = false)
@@ -864,10 +941,16 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{destinationBucket:.+}/**",
-      method = RequestMethod.PUT,
       headers = {
           COPY_SOURCE
-      }, produces = "application/xml")
+      },
+      params = {
+          NOT_UPLOAD_ID
+      },
+      method = RequestMethod.PUT,
+      produces = {
+          APPLICATION_XML_VALUE
+      })
   public ResponseEntity<CopyObjectResult> copyObject(@PathVariable final String destinationBucket,
       @RequestHeader(value = COPY_SOURCE) final ObjectRef objectRef,
       @RequestHeader(value = METADATA_DIRECTIVE,
@@ -917,9 +1000,13 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = "uploads",
+      params = {
+          UPLOADS
+      },
       method = RequestMethod.POST,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      })
   public ResponseEntity<InitiateMultipartUploadResult> initiateMultipartUpload(
       @PathVariable final String bucketName,
       @RequestHeader(value = SERVER_SIDE_ENCRYPTION, required = false) final String encryption,
@@ -954,9 +1041,13 @@ class FileStoreController {
    */
   @RequestMapping(
       value = "/{bucketName:.+}/**",
-      params = {"uploadId"},
+      params = {
+          UPLOAD_ID
+      },
       method = RequestMethod.POST,
-      produces = "application/xml")
+      produces = {
+          APPLICATION_XML_VALUE
+      })
   public ResponseEntity<CompleteMultipartUploadResult> completeMultipartUpload(
       @PathVariable final String bucketName,
       @RequestParam final String uploadId,
@@ -1112,7 +1203,6 @@ class FileStoreController {
   private boolean isV4ChunkedWithSigningEnabled(final String sha256Header) {
     return sha256Header != null && sha256Header.equals(STREAMING_AWS_4_HMAC_SHA_256_PAYLOAD);
   }
-
 
   private Map<String, String> addOverrideHeaders(final String query) {
     if (isNotBlank(query)) {

--- a/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
@@ -845,7 +845,7 @@ class FileStoreController {
       })
   public ResponseEntity<CopyPartResult> copyObjectPart(
       @RequestHeader(value = COPY_SOURCE) final ObjectRef copySource,
-      @RequestHeader(value = COPY_SOURCE_RANGE) final Range copyRange,
+      @RequestHeader(value = COPY_SOURCE_RANGE, required = false) final Range copyRange,
       @RequestHeader(value = SERVER_SIDE_ENCRYPTION, required = false) final String encryption,
       @RequestHeader(
           value = SERVER_SIDE_ENCRYPTION_AWS_KMS_KEYID,
@@ -859,8 +859,7 @@ class FileStoreController {
     final String destinationFile = filenameFrom(destinationBucket, request);
     final String partEtag = fileStore.copyPart(copySource.getBucket(),
         copySource.getKey(),
-        copyRange.getStart(),
-        copyRange.getEnd(),
+        copyRange,
         partNumber,
         destinationBucket,
         destinationFile,

--- a/server/src/main/java/com/adobe/testing/s3mock/domain/FileStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/domain/FileStore.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright 2017-2021 Adobe.
+ *  Copyright 2017-2022 Adobe.
  *
  *  Licensed under the Apache License, Version 2.0 (the "License");
  *  you may not use this file except in compliance with the License.
@@ -27,6 +27,7 @@ import com.adobe.testing.s3mock.dto.CopyObjectResult;
 import com.adobe.testing.s3mock.dto.MultipartUpload;
 import com.adobe.testing.s3mock.dto.Owner;
 import com.adobe.testing.s3mock.dto.Part;
+import com.adobe.testing.s3mock.dto.Range;
 import com.adobe.testing.s3mock.util.AwsChunkDecodingInputStream;
 import com.adobe.testing.s3mock.util.HashUtil;
 import com.fasterxml.jackson.databind.ObjectMapper;
@@ -1137,8 +1138,7 @@ public class FileStore {
    *
    * @param bucket The source Bucket.
    * @param key Identifies the S3 Object.
-   * @param from Byte range form.
-   * @param to Byte range to.
+   * @param copyRange Byte range to copy. Optional.
    * @param partNumber The part to copy.
    * @param destinationBucket The Bucket the target file (will) reside in.
    * @param destinationFilename The target file.
@@ -1150,8 +1150,7 @@ public class FileStore {
    */
   public String copyPart(final String bucket,
       final String key,
-      final long from,
-      final long to,
+      final Range copyRange,
       final String partNumber,
       final String destinationBucket,
       final String destinationFilename,
@@ -1162,16 +1161,20 @@ public class FileStore {
     final File targetPartFile =
         ensurePartFile(partNumber, destinationBucket, destinationFilename, uploadId);
 
-    return copyPart(bucket, key, from, to, targetPartFile);
+    return copyPart(bucket, key, copyRange, targetPartFile);
   }
 
   private String copyPart(final String bucket,
       final String key,
-      final long from,
-      final long to,
+      final Range copyRange,
       final File partFile) throws IOException {
-    final long len = to - from + 1;
+    long from = 0;
     final S3Object s3Object = resolveS3Object(bucket, key);
+    long len = s3Object.getDataFile().length();
+    if (copyRange != null) {
+      from = copyRange.getStart();
+      len = copyRange.getEnd() - copyRange.getStart() + 1;
+    }
 
     try (final InputStream sourceStream = FileUtils.openInputStream(s3Object.getDataFile());
         final OutputStream targetStream = new FileOutputStream(partFile)) {

--- a/server/src/main/java/com/adobe/testing/s3mock/util/AwsHttpParameters.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/util/AwsHttpParameters.java
@@ -1,0 +1,42 @@
+/*
+ *  Copyright 2017-2022 Adobe.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.adobe.testing.s3mock.util;
+
+/**
+ * Holds Parameters used in HTTP requests from AWS S3 Client.
+ */
+public class AwsHttpParameters {
+
+  private static final String NOT = "!";
+
+  public static final String CONTINUATION_TOKEN = "continuation-token";
+  public static final String DELETE = "delete";
+  public static final String ENCODING_TYPE = "encoding-type";
+  public static final String LIST_TYPE_V2 = "list-type=2";
+  public static final String MAX_KEYS = "max-keys";
+  public static final String PART_NUMBER = "partNumber";
+  public static final String START_AFTER = "start-after";
+  public static final String TAGGING = "tagging";
+  public static final String UPLOADS = "uploads";
+
+  public static final String UPLOAD_ID = "uploadId";
+  public static final String NOT_UPLOAD_ID = NOT + UPLOAD_ID;
+
+  private AwsHttpParameters() {
+    // empty private constructor
+  }
+}

--- a/server/src/test/java/com/adobe/testing/s3mock/domain/FileStoreTest.java
+++ b/server/src/test/java/com/adobe/testing/s3mock/domain/FileStoreTest.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright 2017-2021 Adobe.
+ *  Copyright 2017-2022 Adobe.
  *
  *  Licensed under the Apache License, Version 2.0 (the "License");
  *  you may not use this file except in compliance with the License.
@@ -25,6 +25,7 @@ import static org.assertj.core.util.Files.contentOf;
 import com.adobe.testing.s3mock.dto.MultipartUpload;
 import com.adobe.testing.s3mock.dto.Owner;
 import com.adobe.testing.s3mock.dto.Part;
+import com.adobe.testing.s3mock.dto.Range;
 import com.adobe.testing.s3mock.util.HashUtil;
 import java.io.ByteArrayInputStream;
 import java.io.File;
@@ -720,8 +721,37 @@ public class FileStoreTest {
     fileStore.prepareMultipartUpload(TEST_BUCKET_NAME, targetFile, DEFAULT_CONTENT_TYPE,
         ENCODING_GZIP, uploadId, TEST_OWNER, TEST_OWNER);
 
+    Range range = new Range(0, contentBytes.length);
     fileStore.copyPart(
-        TEST_BUCKET_NAME, sourceFile, 0, contentBytes.length, partNumber,
+        TEST_BUCKET_NAME, sourceFile, range, partNumber,
+        TEST_BUCKET_NAME, targetFile, uploadId);
+
+    assertThat(
+        Paths.get(rootFolder.getAbsolutePath(), TEST_BUCKET_NAME, targetFile, uploadId,
+                partNumber + ".part")
+            .toFile()
+            .exists()).as("Part does not exist!").isTrue();
+  }
+
+  @Test
+  public void copyPartNoRange() throws Exception {
+
+    final String sourceFile = UUID.randomUUID().toString();
+    final String uploadId = UUID.randomUUID().toString();
+
+    final String targetFile = UUID.randomUUID().toString();
+    final String partNumber = "1";
+
+    final byte[] contentBytes = UUID.randomUUID().toString().getBytes();
+    fileStore.putS3Object(TEST_BUCKET_NAME, sourceFile, DEFAULT_CONTENT_TYPE, ENCODING_GZIP,
+        new ByteArrayInputStream(contentBytes), false);
+
+    fileStore.prepareMultipartUpload(TEST_BUCKET_NAME, targetFile, DEFAULT_CONTENT_TYPE,
+        ENCODING_GZIP, uploadId, TEST_OWNER, TEST_OWNER);
+
+    Range range = null;
+    fileStore.copyPart(
+        TEST_BUCKET_NAME, sourceFile, range, partNumber,
         TEST_BUCKET_NAME, targetFile, uploadId);
 
     assertThat(
@@ -733,9 +763,10 @@ public class FileStoreTest {
 
   @Test
   public void missingUploadPreparation() {
+    Range range = new Range(0, 0);
     IllegalStateException e = Assertions.assertThrows(IllegalStateException.class, () ->
         fileStore.copyPart(
-            TEST_BUCKET_NAME, UUID.randomUUID().toString(), 0, 0, "1",
+            TEST_BUCKET_NAME, UUID.randomUUID().toString(), range, "1",
             TEST_BUCKET_NAME, UUID.randomUUID().toString(), UUID.randomUUID().toString())
     );
 

